run/service-auth/app.py

```diff
@@ -12,30 +12,133 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+# [START auth_validate_and_decode_bearer_token_on_flask]
+# [START cloudrun_service_to_service_receive]
+"""Demonstrates how to receive authenticated service-to-service requests
+on a Cloud Run Service.
+"""
+
 from http import HTTPStatus
 import os
+from typing import Optional
 
 from flask import Flask, request
 
-from receive import receive_request_and_parse_auth_header
+from google.auth.exceptions import GoogleAuthError
+from google.auth.transport import requests
+from google.cloud import run_v2
+from google.oauth2 import id_token
 
 app = Flask(__name__)
 
 
+def get_service_url(full_service_name: str) -> str:
+    """Retrieves the primary URL for a given Cloud Run service.
+
+    Args:
+        full_service_name: The fully qualified name of the Cloud Run service.
+            The expected format is
+            "projects/PROJECT_ID/locations/REGION/services/SERVICE_NAME".
+
+    Returns:
+        The base URL of the specified Cloud Run service.
+    """
+    client = run_v2.ServicesClient()
+
+    service_request = run_v2.GetServiceRequest(
+        name=full_service_name,
+    )
+
+    service_info = client.get_service(request=service_request)
+
+    # Get the Deterministic URL for this Service and remove any trailing slash.
+    # https://cloud.google.com/run/docs/triggering/https-request#deterministic
+    service_url = service_info.urls[0].rstrip('/')
+
+    return service_url
+
+
+def parse_auth_header(auth_header: str) -> Optional[str]:
+    """Parse the authorization header, validate and decode the Bearer token.
+
+    Args:
+        auth_header: Raw HTTP header with a Bearer token.
+
+    Returns:
+        A string containing the email from the token.
+        None if the token is invalid or the email can't be retrieved.
+    """
+
+    # Split the auth type and value from the header.
+    try:
+        auth_type, creds = auth_header.split(" ", 1)
+    except ValueError:
+        print("Malformed Authorization header.")
+        return None
+
+    # Get the full service name from the environment variable
+    # set at the time of deployment.
+    # Format: "projects/PROJECT_ID/locations/REGION/services/SERVICE_NAME"
+    full_service_name = os.getenv("FULL_SERVICE_NAME")
```

Comment:
It looks like you've changed the previous method to go from using the metadata service for the entire logic, to having the full resource name supplied as an environment variable. 

It would be simpler just to get the user to add the service URL as a environment variable. Since they're adding something, add the variable that's directly required. 

The Django tutorial uses a similar method https://cloud.google.com/python/django/run#deploy